core/system/src/page_allocator.c

```diff
@@ -0,0 +1,287 @@
+
+static inline int vmpu_is_region_size_valid(uint32_t size)
+{
+    /* Minimal 32B alignment */
+    return ((size & ~31) == size);
+}
+static inline uint32_t vmpu_round_up_region(uint32_t addr, uint32_t size)
+{
+    if (!vmpu_is_region_size_valid(size)) {
+        return 0;
+    }
+    const uint32_t mask = size - 1;
+    return (addr + mask) & ~mask;
+}
+static inline uint32_t page_table_read(uint32_t addr)
+{
+    return *((uint32_t *) addr);
+}
+static inline void page_table_write(uint32_t addr, uint32_t data)
+{
+    *((uint32_t *) addr) = data;
+}
+
+#endif /* defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1) */
+
+/* By default a maximum of 16 pages are allowed. This can only
+ * be decided by the porting engineer for the current platform! */
+#ifndef UVISOR_PAGE_TABLE_MAX_COUNT
+#define UVISOR_PAGE_TABLE_MAX_COUNT ((size_t) 16)
+#endif
+/* The page box_id is the box id which is 8bit large  */
+typedef uint8_t page_owner_t;
+/* Maps the page to the owning box handle */
+static page_owner_t g_page_owner_table[UVISOR_PAGE_TABLE_MAX_COUNT];
+/* define a unused value for the page table */
+#define UVISOR_PAGE_UNUSED ((page_owner_t) (-1))
+
+/* contains the configured page size */
+static uint32_t g_page_size;
+/* Points to the beginning of the page heap */
+static const void * g_page_heap_start;
+/* Points to the end of the page heap */
+static const void * g_page_heap_end;
+/* Contains the number of free pages */
+static uint8_t g_page_count_free;
+/* Contains the total number of available pages */
+static uint8_t g_page_count_total;
+
+
+void page_allocator_init(void * const heap_start, void * const heap_end, const uint32_t * const page_size)
+{
+    /* Make sure the UVISOR_PAGE_UNUSED is definitely NOT a valid box id! */
+    if (vmpu_is_box_id_valid(UVISOR_PAGE_UNUSED)) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "UVISOR_PAGE_UNUSED (%u) must not be a valid box id!\n",
+            UVISOR_PAGE_UNUSED);
+    }
+    if (!page_size || !vmpu_public_flash_addr((uint32_t) page_size)) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "Page size pointer (0x%08x) is not in flash memory.\n",
+            (unsigned int) page_size);
+    }
+    if (!heap_start || !vmpu_sram_addr((uint32_t) heap_start)) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "Page heap start pointer (0x%08x) is not in sram memory.\n",
+            (unsigned int) heap_start);
+    }
+    if (!heap_end || !vmpu_sram_addr((uint32_t) heap_end)) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "Page heap end pointer (0x%08x) is not in sram memory.\n",
+            (unsigned int) heap_end);
+    }
+    if (heap_end < heap_start) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "Page heap end pointer (0x%08x) is smaller than heap start pointer (0x%08x).\n",
+            (unsigned int) heap_end,
+            (unsigned int) heap_start);
+    }
+    if ((*page_size < 1024) || !vmpu_is_region_size_valid(*page_size)) {
```

Comment:
What went into the decision to make 1024 the minimum page size? Add a comment in the code.